### PR TITLE
Pyramid create for creating pyramids with custom funcs

### PR DESCRIPTION
**Pull request by ahuang11**
Closes https://github.com/carbonplan/ndpyramid/issues/94

Adapted pyramid_coarsen from ndpyramid [for a project](https://github.com/holoviz-topics/neuro/issues/87#issuecomment-1992668594), but I would like to add a bit of functionality back so I can simply import it.

Basic usage:
```python
def sel_coarsen(ds, factor, dims, **kwargs):
    return ds.sel(**{dim: slice(None, None, factor) for dim in dims})

factors = [4, 2, 1]
pyramid = pyramid_create(
    temperature,
    dims=('lat', 'lon'),
    factors=factors,
    boundary='trim',
    func=sel_coarsen,
    method_label=method_label,
    type_label='pick',
)
```

I think it could eventually support https://github.com/carbonplan/ndpyramid/issues/10, but haven't looked at it in detail.

More advanced, applicable use case:
```python
import h5py
import xarray as xr
import dask.array as da
import datatree as dt
from ndpyramid import pyramid_create
from tsdownsample import MinMaxLTTBDownsampler


def _help_downsample(data, time, n_out):
    indices = MinMaxLTTBDownsampler().downsample(time, data, n_out=n_out)
    return data[indices], indices


def apply_downsample(ts_ds, factor, dims):
    dim = dims[0]
    n_out = len(ts_ds["data"]) // factor
    ts_ds_downsampled, indices = xr.apply_ufunc(
        _help_downsample,
        ts_ds["data"],
        ts_ds[dim],
        kwargs=dict(n_out=n_out),
        input_core_dims=[[dim], [dim]],
        output_core_dims=[[dim], ["indices"]],
        exclude_dims=set((dim,)),
        vectorize=True,
        dask="parallelized",
        dask_gufunc_kwargs=dict(output_sizes={dim: n_out, "indices": n_out}),
    )
    ts_ds_downsampled[dim] = ts_ds[dim].isel(time=indices.values[0])
    return ts_ds_downsampled.rename("data")


def build_dataset(f, data_key, dims):
    coords = {f[dim] for dim in dims.values()}
    data = f[data_key]
    ds = xr.DataArray(
        da.from_array(data, name="data", chunks=(data.shape[0], 1)),
        dims=dims,
        coords=coords,
    ).to_dataset()
    return ds


f = h5py.File("allensdk_cache/session_715093703/probe_810755797_lfp.nwb", "r")
ts_ds = build_dataset(
    f,
    "acquisition/probe_810755797_lfp_data/data",
    {
        "time": "acquisition/probe_810755797_lfp_data/timestamps",
        "channel": "acquisition/probe_810755797_lfp_data/electrodes",
    },
).isel(channel=[0, 1, 2, 3, 4])
ts_dt = pyramid_create(
    ts_ds,
    factors=[1, 2],
    dims=["time"],
    func=apply_downsample,
    type_label="pick",
    method_label="pyramid_downsample",
)
ts_dt
```

### CODE REVIEW

File: docs/generate-pyramids.md [hunk 1]
```diff
@@ -23,3 +23,25 @@ pyramid = pyramid_reproject(ds, levels=2)
...
+    return ds.sel(**{dim: slice(None, None, factor) for dim in dims})
+
+factors = [4, 2, 1]
+pyramid = pyramid_create(
+    temperature,
+    dims=('lat', 'lon'),
+    factors=factors,
+    boundary='trim',
+    func=sel_coarsen,
+    method_label=method_label,
```
Comment (maxrjones) [inline docs/generate-pyramids.md:44]:
```suggestion
    method_label="slice_coarsen",
```

File: docs/generate-pyramids.md [hunk 2]
```diff
@@ -23,3 +23,25 @@ pyramid = pyramid_reproject(ds, levels=2)
...
 ```
+
+There's also `pyramid_create`--a more versatile alternative to `pyramid_coarsen`.
+
+This function accepts a custom function with the signature: `ds`, `factor`, `dims`.
+
+Here, the `sel_coarsen` function uses `ds.sel` to perform coarsening:
+
+```python
+def sel_coarsen(ds, factor, dims, **kwargs):
```
Comment (maxrjones) [inline docs/generate-pyramids.md:34]:
```suggestion
from ndpyramid import pyramid_create

def sel_coarsen(ds, factor, dims, **kwargs):
```

File: docs/generate-pyramids.md [hunk 3]
```diff
@@ -23,3 +23,25 @@ pyramid = pyramid_reproject(ds, levels=2)
 # write the pyramid to zarr
 pyramid.to_zarr('./path/to/write')
 ```
+
+There's also `pyramid_create`--a more versatile alternative to `pyramid_coarsen`.
+
+This function accepts a custom function with the signature: `ds`, `factor`, `dims`.
+
+Here, the `sel_coarsen` function uses `ds.sel` to perform coarsening:
```
Comment (maxrjones) [inline docs/generate-pyramids.md:31]:
```suggestion
There's also `pyramid_create` -- a more versatile alternative to `pyramid_coarsen`.
This function accepts a custom function with the signature: `ds`, `factor`, `dims`.
Here, the `sel_coarsen` function uses `ds.sel` to perform coarsening:
```

**Review by maxrjones — APPROVED** [review]
Thanks @ahuang11, this is nice! I appreciate that you shared your use case to show how this would broaden the usefulness of the library! I made a couple small docs suggestions. Also, can you please add `pyramid_create` to `docs/api.rst` so that it is rendered in the API docs?

**Review by maxrjones — APPROVED** [review]
Thanks @ahuang11!

**Comment by ahuang11** [comment]
Thanks for merging! I was wondering when you envision the next release to be?

**Comment by maxrjones** [comment]
> Thanks for merging! I was wondering when you envision the next release to be?

I can put out a release this week

**Comment by ahuang11** [comment]
I was wondering if you had a chance to put out a new release yet?

**Comment by maxrjones** [comment]
> I was wondering if you had a chance to put out a new release yet?

Thanks for the reminder - just put out a new release

**Pull request merged**